services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java

```diff
@@ -0,0 +1,76 @@
+package com.linkedin.openhouse.tables.api.validator.impl;
+
+import com.linkedin.openhouse.common.api.spec.TableUri;
+import com.linkedin.openhouse.tables.api.spec.v0.request.components.History;
+import com.linkedin.openhouse.tables.api.spec.v0.request.components.TimePartitionSpec;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+public class HistoryPolicySpecValidator {
+
+  private String failureMessage = "";
+  private String errorField = "";
+
+  protected boolean validate(History history, TableUri tableUri) {
+    if (history != null) {
+      if (history.getMaxAge() <= 0 && history.getVersions() <= 0) {
+        failureMessage =
+            String.format(
+                "Must define either a time based retention or count based retention for snapshots in table %s",
+                tableUri);
+        return false;
+      }
+
+      if (history.getGranularity() == null && history.getMaxAge() > 0
+          || history.getGranularity() != null && history.getMaxAge() <= 0) {
+        failureMessage =
+            String.format(
+                "Incorrect maxAge specified. history.maxAge must be defined together with history.granularity for table %s",
+                tableUri);
+        return false;
+      }
+
+      if (!validateHistoryConfigMaximums(history)) {
+        failureMessage =
+            String.format(
+                "History for the table [%s] cannot exceed 3 days or 100 versions maximum",
+                tableUri);
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Validate that the amount of time to retain snapshots does not exceed either 3 days or 100
+   * versions
+   *
+   * @param history
+   * @return
+   */
+  protected boolean validateHistoryConfigMaximums(History history) {
```

Comment:
shall we also consider minimum version ? Let me know if this is being discussed somewhere else

Reply:
I'll modify this comment good catch